pkg/controller/nginxingresscontroller/nginxingresscontroller_controller.go

```diff
@@ -309,6 +310,25 @@ func (r *ReconcileNginxIngressController) checkPrerequisites(reqLogger logr.Logg
 		reqLogger.Info("ConfigMap created", "ConfigMap.Namespace", cm.Namespace, "ConfigMap.Name", cm.Name)
 	}
 
+	// IngressClass is available from k8s 1.18+
+	minVersion, _ := version.ParseGeneric("v1.18.0")
+	if RunningK8sVersion.AtLeast(minVersion) {
+		if instance.Spec.IngressClass == "" {
+			instance.Spec.IngressClass = "nginx"
+			reqLogger.Info("Warning! IngressClass not set, using default", "IngressClass.Name", instance.Spec.IngressClass)
+		}
+		ic := ingressClassForNginxIngressController(instance)
+
+		err, existed = r.createIfNotExists(ic)
```

Comment:
For similar objects (ConfigMap, DefaultSecret), we set the controller reference.

For example, for the default secret, we have `err = controllerutil.SetControllerReference(instance, secret, r.scheme)`. This sets the `ownerReferences` of the secret to the NginxIngressController resource. This means that if IngressClass resource is deleted, Kubernetes will remove the secret. See also https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/

Could you do the same for IngressClass resource? So that if `NginxIngressController` is removed, the IngressClass resource is deleted as well

Reply:
The IngressClass is at the cluster level and can't have an `ownerReferences` that's why I didn't use that. I'm trying to figure out if I need to use a new reconciler...

Reply:
I still haven't figured this out. Any ideas @Dean-Coakley @pleshakov ?

Reply:
@lucacome 

how about removing Ingress Class in `finalizeNginxIngressController` ? If  an `NginxIngressController` is marked for deletion, you remove the Ingress Class resource.

Reply:
I can't get it to work manually either. I'd say to add it as tech debt and just merge this for now. Also, I noticed that we don't delete other resources, like CRDs.
Maybe we can take another look when moving to the operator SDK v0.19.0

Reply:
what is an impact of that? 
how can a problem be mitigated?
will it be fixed before a new release?  

CRDs are a bit different, because they are share among all KICs.

Reply:
Well, we create a new `IngressClass` resource for the `ingressClass` specified in `NginxIngressController` if it doesn't exist already. Since they have the controller set to `"nginx.org/ingress-controller"` they're only for our IC.

> how can a problem be mitigated?

not sure I understand what you mean. Are you asking how the user can fix this? If so they can always remove the IngressClass manually after deleting the IC.

> will it be fixed before a new release?

Yes that's the plan, to update the SDK and fix this before release.

Reply:
thanks

> not sure I understand what you mean. Are you asking how the user can fix this? If so they can always remove the IngressClass manually after deleting the IC.

yep, if we were to release something with a problem, some documentation for mitigating the problem would be necessary